### ISSUE TITLE
[BO] Partial refund generates wrong Orderslip

### ISSUE THREAD

**Issue by ChillCode** [PrestaShop/PrestaShop#32762]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When a partial refund is made on a shop set to show prices without taxes the generated Orderslip has wrong taxes.

### Expected behavior

There is a partial column on ps_order_slip that if set to 1 on Orderslip creation then classes/pdf/HTMLTemplateOrderSlip.php takes that in count and uses the right taxes from db.

### Steps to reproduce

1. Install a clean PrestaShop 8.0.4.
2. [BO] Go to `Shop Parameters->Customer Settings->Groups` and change the `Price display method` to `Tax Excluded` of all three groups.
3. [BO] Go to `International->Taxes` and edit TVA FR 20% and set 21 or 22 as `Rate`.
4. [FO] Place an order with two products with different prices  x10 quantity each product. Total order quantity 20.
3. [BO] Open the placed order and mark it as paid.
4. [BO] Partial refund 5 units of one of the products and make sure Orderslip generation is checked.
6. [BO] Download the generated Orderslip and calculate/check the taxes.
7. [BO] Make Partial Refunds till the whole order is refunded and sum all totals and compare it to the invoice.

Invoice:

Total 498,98€

[IN000007.pdf](https://github.com/PrestaShop/PrestaShop/files/11733818/IN000007.pdf)

Total Refunded: 499€

[000010.pdf](https://github.com/PrestaShop/PrestaShop/files/11733832/000010.pdf)
[000011.pdf](https://github.com/PrestaShop/PrestaShop/files/11733833/000011.pdf)
[000012.pdf](https://github.com/PrestaShop/PrestaShop/files/11733834/000012.pdf)

In some configurations the difference is not 2 cents, but it would require to add discounts by quantities.

After I set partial column in ps_order_slip table manually to 1 or just remove some code from HTMLTemplateOrderSlip.php the result is correct:

Total Refunded: 498,98€

[000010_1.pdf](https://github.com/PrestaShop/PrestaShop/files/11733876/000010_1.pdf)
[000011_1.pdf](https://github.com/PrestaShop/PrestaShop/files/11733877/000011_1.pdf)
[000012_1.pdf](https://github.com/PrestaShop/PrestaShop/files/11733878/000012_1.pdf)

### PrestaShop version(s) where the bug happened

8.0.4

### PHP version(s) where the bug happened

8.1

**Comment by AchrafKchaou1991** [comment]
Hello @ChillCode 
I reproduce the issue with PrestaShop versions 8.0.4 and 1.7.5.2 , I'll add this to the backlog so it can be fixed.
Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you.
If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

![Capture d’écran du 2023-06-01 14-31-32](https://github.com/PrestaShop/PrestaShop/assets/131356155/27ced3eb-9dfb-4c8f-b67e-ce7d98ce22bd)

Thank you

**Comment by Misty-94** [comment]
Hello,

What's wrong with your example ?

Total with tax : 14.28
Total without tax : 14.28 / 1.20 = 11.90
Tax : 14.28 - 11.90 = 2.38

For me it's OK ?

**Comment by Misty-94** [comment]
The problem is the minus sign maybe ?

**Comment by Misty-94** [comment]
Even if we calculate with the minus signs :

Total with tax : -14.28
Total without tax : -14.28 / 1.20 = -11.90
Tax : -14.28 - (-11.90) = -14.28 + 11.90 = -2.38

**Comment by Misty-94** [comment]
Should product prices be without taxes instead ? and then the first total should be without taxes ?

**Comment by ChillCode** [comment]
Sorry, when the user complained didn't remember that he has Quantities Discounts, the Visitor/Guess/Customer Group are configured to show prices without taxes and taxes are 21%.

I'll update my issue adding more steps.

Btw, here is the result on a clean install:

[000005.pdf](https://github.com/PrestaShop/PrestaShop/files/11650282/000005.pdf)

**Comment by RosaBenouamer** [comment]
I am lookink on the new pdf. 
The taxes seems to be good
-10,46 * 0,21 * 20 = -44
The taxes excl seems to be good
-10,46 * 20 = -209,2
and the tax included too
-209,2 - 44 = 253,2
Isn't it ?
@ChillCode or perhaps I don't understand your question ?
@Zoefinders perhaps waiting the response of the author before put the ticket in Maxi Kanban/Ready

**Comment by ChillCode** [comment]
Hi

When we enter data to ERP software it uses the base to calculate taxes, so when we enter 209,2 ( * 1,21) the result is 253,13€

If we manually set the value to 1 in partial column of ps_order_slip it gives the correct result:

[000005_1.pdf](https://github.com/PrestaShop/PrestaShop/files/11651760/000005_1.pdf)

But watching through the code that column is never set to 1, there is a method that's sets it but is never called anywhere.

**Comment by ChillCode** [comment]
Also this calc is not correct:

> -10,46 * 0,21 * 20 = -44

-10,46 * 0,21 * 20 = -43.93

**Comment by RosaBenouamer** [comment]
Looking at your first pdf I have read 44 so I make a round supposing a round mode :-) Of course without round the result is 43,93 :-)

**Comment by RosaBenouamer** [comment]
> Hi
> 
> When we enter data to ERP software it uses the base to calculate taxes, so when we enter 209,2 ( * 1,21) the result is 253,13€
> 
> If we manually set the value to 1 in partial column of ps_order_slip it gives the correct result:
> 
> [000005_1.pdf](https://github.com/PrestaShop/PrestaShop/files/11651760/000005_1.pdf)
> 
> But watching through the code that column is never set to 1, there is a method that's sets it but is never called anywhere.

What is the correspondance in BO of "ps_order_slip" ? I suppose ps_order_slip is a technical name ?

**Comment by WahbiPS** [comment]
Blokced for the moment, @Zoefinders checks with @RosaBenouamer

**Comment by RosaBenouamer** [comment]
Hello @ChillCode,
We have tested with @paulnoelcholot and have the good result.
Can you precise you parameter in :
- Interational > Currencies > [Your currency] > Decimals
- Shop Parameters > General > Round mode
- Shop Parameters > General > Round type
![image_720](https://github.com/PrestaShop/PrestaShop/assets/124665010/a5fd6891-72da-407d-a08d-b4e176e83663)

**Comment by RosaBenouamer** [comment]
@ChillCode I notice in you first pdf unit price and total is (Tax incl.) in contrary of the second pdf. How do you change this parameter ? PS : @paulnoelcholot

**Comment by ChillCode** [comment]
Sorry for not answering before but wanted to make some more tests.

@RosaBenouamer To have prices without taxes I go to `Shop Parameters->Groups` and edit all three groups to change `Price display method`  to Tax ecluded.

**Comment by RosaBenouamer** [comment]
@paulnoelcholot could you check please ?

**Comment by RosaBenouamer** [comment]
@ChillCode sorry I was sick. @paulnoelcholot did you succeed to reproduce with the parameters ?

**Comment by RosaBenouamer** [comment]
Hello @ChillCode :-) 
Could you please give us all the parameters with wich you encountered a problem :
If you could make a video and complete description above with the max of information 
- wich article, prices
- quantity
- Interational > Currencies > [Your currency] > Decimals
- Shop Parameters > General > Round mode
- Shop Parameters > General > Round type
- ...
- on wich articles you ask a partial refund with quantity and amount...
- expected behavior
- and the result you obtain

**Comment by ChillCode** [comment]
Hi

Updated the steps.

It's a clean install that I only change the groups to display prices without taxes and edited French tax to 22 (I wanted to try with 22) so the rounding error appears easily.

My first report was on a site that had also Discount prices by quantity and the rounding errors are even bigger but would require to edit the products to reproduce so now if you follow the steps you should see the rounding errors.

**Comment by paulnoelcholot** [comment]
Hello @ChillCode,

Thanks for your update. Now I understand and I reproduce the issue.

Invoice : 
![image](https://github.com/PrestaShop/PrestaShop/assets/112863083/37e0fe30-7df3-4ce1-b976-6afa1ea787ed)

Refund 1 : 
![image](https://github.com/PrestaShop/PrestaShop/assets/112863083/5c5c9d4b-16fb-4346-9274-c3423cad804c)

Refund 2 : 
![image](https://github.com/PrestaShop/PrestaShop/assets/112863083/31e6eebf-b928-4d9e-a085-e54c31f858e3)

Refund 3 : 
![image](https://github.com/PrestaShop/PrestaShop/assets/112863083/6bff8da0-adc3-44c3-a4d4-e62a0f17ae39)

Invoice = 498.98€
Refund 1 + 2 + 3 = 176.90 + 145.20 + 176.90 = 499.00 €

@RosaBenouamer can you confirm that this is an issue?

**Comment by RosaBenouamer** [comment]
yes @paulnoelcholot and as we have seen with only one refund the number 2 we have the issue

**Comment by RosaBenouamer** [comment]
@MatShir the problem is on the refund 2.
You have a Tax at 22%
Expected 0.22 * 119 = 26,18 
but we have 26.20
In the two cases you have two numbers after the coma so no rounded expected.
